File: python/Heap.py
class MinHeap:
    def __init__(self):
        self.Heap=[]
    def parent(self,i):
        return (i-1)//2
    def insertKey(self,k):
        self.Heap.append(k)
        i=len(self.Heap)-1
        while i!=0 and self.Heap[self.parent(i)]>self.heap[i]:
            self.heap[i],self.heap[self.parent(i)]=self.heap[self.parent(i)],self.heap[i]
            i=self.parent(i)


    def heapify(self,i):
        smallest =i 
        left=2*i+1
        right=2*i+2
        if left<len(self.heap) and self.heap[left]<self.heap[i]:
            smallest=left
        if right <len(self.heap) and self.heap[right]<self.heap[i]:
            smallest=right
        if smallest !=i:
            self.heap[i],self.heap[smallest]= self.heap[smallest],self.heap[i]
            self.heapify(smallest)


    def extractMin(self):
        if len(self.heap) ==0:
            return float("inf")
        if len(self.heap) ==1:
            return self.heap.pop()
        root=self.heap[0]
        self.heap[0]=self.heap.pop()
        self.heapify(0)
        return root


    def getMin(self):
        return self.heap[0] if self.heap else None


    def display(self):
        return self.Heap    




class MaxHeap:
    def __init__(self):
        self.heap = []

    def parent(self, i):
        return (i - 1) // 2

    def insertKey(self, k):
        self.heap.append(k)
        i = len(self.heap) - 1
        while i != 0 and self.heap[self.parent(i)] < self.heap[i]:
            self.heap[i], self.heap[self.parent(i)] = self.heap[self.parent(i)], self.heap[i]
            i = self.parent(i)

    def heapify(self, i):
        largest = i
        left = 2 * i + 1
        right = 2 * i + 2

        if left < len(self.heap) and self.heap[left] > self.heap[i]:
            largest = left
        if right < len(self.heap) and self.heap[right] > self.heap[largest]:
            largest = right

        if largest != i:
            self.heap[i], self.heap[largest] = self.heap[largest], self.heap[i]
            self.heapify(largest)

    def extractMax(self):
        if len(self.heap) == 0:
            return float("-inf")
        if len(self.heap) == 1:
            return self.heap.pop()
        root = self.heap[0]
        self.heap[0] = self.heap.pop()
        self.heapify(0)
        return root

    def getMax(self):
        return self.heap[0] if self.heap else None

    def display(self):
        return self.heap


def main():
    min_heap = MinHeap()
    max_heap = MaxHeap()

    while True:
        print("\nOptions:")
        print("1. Insert into Min Heap")
        print("2. Delete Min from Min Heap")
        print("3. Display Min Heap")
        print("4. Insert into Max Heap")
        print("5. Delete Max from Max Heap")
        print("6. Display Max Heap")
        print("7. Exit")

        choice = int(input("Enter your choice: "))

        if choice == 1:
            element = int(input("Enter the element to insert in Min Heap: "))
            min_heap.insertKey(element)
        elif choice == 2:
            print("Minimum Element Removed:", min_heap.extractMin())
        elif choice == 3:
            print("Min Heap:", min_heap.display())
        elif choice == 4:
            element = int(input("Enter the element to insert in Max Heap: "))
            max_heap.insertKey(element)
        elif choice == 5:
            print("Maximum Element Removed:", max_heap.extractMax())
        elif choice == 6:
            print("Max Heap:", max_heap.display())
        elif choice == 7:
            break
        else:
            print("Invalid choice. Please enter again.")

if __name__ == "__main__":
    main()

File: R/Heap.r
MinHeap <- setRefClass("MinHeap",
    fields = list(heap = "numeric"),
    methods = list(
        insert = function(key) {
            heap <<- c(heap, key)
            self$heapifyUp(length(heap))
        },
        extractMin = function() {
            if (length(heap) == 0) {
                return(NA)
            }
            root <- heap[1]
            heap[1] <- heap[length(heap)]
            heap <<- heap[-length(heap)]
            self$heapifyDown(1)
            return(root)
        },
        heapifyUp = function(i) {
            while (i > 1 && heap[i] < heap[i %/% 2]) {
                temp <- heap[i]
                heap[i] <- heap[i %/% 2]
                heap[i %/% 2] <- temp
                i <- i %/% 2
            }
        },
        heapifyDown = function(i) {
            smallest <- i
            left <- 2 * i
            right <- 2 * i + 1
            if (left <= length(heap) && heap[left] < heap[smallest]) {
                smallest <- left
            }
            if (right <= length(heap) && heap[right] < heap[smallest]) {
                smallest <- right
            }
            if (smallest != i) {
                temp <- heap[i]
                heap[i] <- heap[smallest]
                heap[smallest] <- temp
                self$heapifyDown(smallest)
            }
        }
    )
)
MaxHeap <- setRefClass("MaxHeap",
    fields = list(heap = "numeric"),
    methods = list(
        insert = function(key) {
            heap <<- c(heap, key)
            self$heapifyUp(length(heap))
        },
        extractMax = function() {
            if (length(heap) == 0) {
                return(NA)
            }
            root <- heap[1]
            heap[1] <- heap[length(heap)]
            heap <<- heap[-length(heap)]
            self$heapifyDown(1)
            return(root)
        },
        heapifyUp = function(i) {
            while (i > 1 && heap[i] > heap[i %/% 2]) {
                temp <- heap[i]
                heap[i] <- heap[i %/% 2]
                heap[i %/% 2] <- temp
                i <- i %/% 2
            }
        },
        heapifyDown = function(i) {
            largest <- i
            left <- 2 * i
            right <- 2 * i + 1
            if (left <= length(heap) && heap[left] > heap[largest]) {
                largest <- left
            }
            if (right <= length(heap) && heap[right] > heap[largest]) {
                largest <- right
            }
            if (largest != i) {
                temp <- heap[i]
                heap[i] <- heap[largest]
                heap[largest] <- temp
                self$heapifyDown(largest)
            }
        }
    )
)
minH <- MinHeap$new()
minH$insert(3)
minH$insert(2)
minH$insert(15)
print(minH$extractMin()) # Should return 2

maxH <- MaxHeap$new()
maxH$insert(3)
maxH$insert(2)
maxH$insert(15)
print(maxH$extractMax()) # Should return 15
